xip/xipether.c

```diff
@@ -0,0 +1,449 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <sys/types.h>
+#include <asm/byteorder.h>
+#include <asm-generic/errno-base.h>
+#include <net/xia_fib.h>
+#include <xia_socket.h>
+#include <sys/socket.h>
+#include <linux/netdevice.h>
+
+#include "xip_common.h"
+#include "utils.h"
+#include "ll_map.h"
+#include "libnetlink.h"
+#include "xiart.h"
+
+#define IFINDEX_STR_SIZE 8
+#define XIA_LLADDR_LEN 12
+#define APPEND_ZEROES 20
+
+static int usage(void)
+{
+	fprintf(stderr,
+"Usage: xip ether { addinterface | delinterface } dev DEV\n"
+"       xip ether { addneigh | delneigh } lladdr LLADDR dev DEV\n"
+"       xip ether show { interfaces | neighs }\n"
+"where  LLADDR := HEXDIGIT{1,2} (':' HEXDIGIT{1,2})*\n"
+"       DEV := STRING NUMBER\n");
+	return -1;
+}
+
+static int form_ether_xid(unsigned oif, unsigned char *lladdr, unsigned tlen,
+			  char *id)
+{
+	char bytes[4];
+	unsigned append = APPEND_ZEROES;
+	int rc;
+
+	bytes[0] = (oif >> 24) & 0xFF;
+	bytes[1] = (oif >> 16) & 0xFF;
+	bytes[2] = (oif >> 8) & 0xFF;
+	bytes[3] = oif & 0xFF;
+	rc = snprintf(id, (tlen + 1),
+			"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%0*hu",
+			(unsigned char)bytes[0], (unsigned char)bytes[1],
+			(unsigned char)bytes[2], (unsigned char)bytes[3],
+			lladdr[0], lladdr[1], lladdr[2], lladdr[3], lladdr[4],
+			lladdr[5], append, 0);
+	return rc;
```

Comment:
The return value of `snprintf()` (which is assigned to `rc` here) is a little bit complicated, so we need to adjust what we return from `form_ether_xid()` in order to indicate failure. Instead of just returning `rc`, let's make it:

    if (rc <= 0 || rc >= tlen)
        return -1;
    return 0;